### PR TITLE
handler: add Ref check to WebhookHandler

### LINKED ISSUE #252: Webhook code should filter for a ref/branch name match before kicking off
**What version of eunomia are you using?**

eunomia version:
v.0.1.0
**Does this issue reproduce with the latest release?**
yes

**What operating system and processor architecture are you using (`kubectl version`)?**

<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version

</pre></details>

**What did you do?**
done a git push to a branch ref that is not configured in a GitOpsConfig CR

**What did you expect to see?**
Eunomia should not kickoff any jobs when the webhook payload indicates 'git push' was done to a branch name different to a ref name specified in the GitOpsConfig CR deployed in the cluster

**What did you see instead?**
Eunomia kicks-off a job/pod upon receiving any 'git push' webhook, no matter if the GitOpsConfig CR specifed ref matches the one from webhook

### PR DESCRIPTION
**Pull request by rhalat**
**Description**

GitHub webhook is triggered on all Push events and without checking 
branch in WebhookHandler new job is created every time there is push to 
repo, no matter what branch. In this PR I added check, so job is created 
only when URI and Ref matches in push event and GitOpsConfig CR.

Additionally it's fixing Travis CI that was failing on hello-world-yaml-cr1 test by increasing timeout.

Fixes #252

**Type of change**

* Bug fix (non-breaking change which fixes an issue)

**Checklist**

- [x] Unit tests and e2e tests updated
- [x] Documentation updated

### CODE REVIEW

File: pkg/handler/handler.go [hunk 1]
```diff
@@ -117,8 +117,11 @@ func WebhookHandler(w http.ResponseWriter, r *http.Request, reconciler gitopscon
 	log.Info("webhook handling concluded correctly")
 }
 
-func repoURLMatch(instance *gitopsv1alpha1.GitOpsConfig, event *github.PushEvent) bool {
-	return strings.Contains(instance.Spec.TemplateSource.URI, *event.Repo.FullName) || strings.Contains(instance.Spec.ParameterSource.URI, *event.Repo.FullName)
+func repoURLAndRefMatch(instance *gitopsv1alpha1.GitOpsConfig, event *github.PushEvent) bool {
```
Comment (akavel) [inline pkg/handler/handler.go]: Please add a unit test for this func, with at least 1 positive and 1 negative test case. Some helpful links for learning how to write unit tests in Go in a standard way:

- https://golang.org/doc/code.html#Testing
- https://dave.cheney.net/2019/05/07/prefer-table-driven-tests

Suggested initial template for the test function:

```
func TestRepoURLAndRefMatch(t *testing.T) {
    tests := []struct{
        comment  string
        instance *gitopsv1alpha1.GitOpsConfig
        event    *github.PushEvent
        want     bool
    }{
        {
             comment: "full match",
             ...
             want: true,
        },
        {
             comment: "parameterSource.Ref differs",
             ...
             want: false,
        },
        {
             comment: "parameterSource.URI differs",
             ...
             want: false,
        },
    }

    for _, tt := range tests {
         result := repoURLAndRefMatch(tt.instance, tt.event)
         if result != tt.want {
             t.Errorf("%q: expected %v, got %v", tt.comment, tt.want, match)
         }
    }
}
```

**Comment by seanmalloy** [comment]
@rhalat what if `.Spec.ParameterSource.Ref` or `.Spec.TemplateSource.Ref` is not a branch? How does this code handle a git tag or git sha1?

**Comment by rhalat** [comment]
@seanmalloy is it supposed to handle something different than branches? When it comes to sha1, I believe it won't trigger job and when it comes to tags we can think of it and I can include it in unit tests, but what in case where branch and tag have same names? I GitOpsConfig CR we are not distinguish them.

**Comment by akavel** [comment]
@seanmalloy @rhalat Maybe we could just tell users to set Ref to `refs/tags/foobar` when they want to trigger on `foobar` tag? I suppose this probably already works now with this PR, and would also help us avoid name conflicts?

**Comment by seanmalloy** [comment]
@rhalat please make the code work with git branches and git tags. We can worry about the git SHA1 use case at some point in the future.

As @akavel mentioned maybe we just need to update the docs for now, so people know how to use the webhook trigger with both branches and tags.

**Comment by rhalat** [comment]
I updated README.md and I'm close to finish with unit tests.

File: pkg/handler/handler_test.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handler
+
+import (
+	gitopsv1alpha1 "github.com/KohlsTechnology/eunomia/pkg/apis/eunomia/v1alpha1"
+	"github.com/google/go-github/github"
+	"testing"
```
Comment (akavel) [inline pkg/handler/handler_test.go]: Please split imports into groups, separated by empty line:

1. standard library imports
2. third-party imports
3. project-internal imports

File: pkg/handler/handler_test.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+			want: true,
+		},
+		{
+			comment: "Ref differs",
+			spec: gitopsv1alpha1.GitOpsConfigSpec{
+				TemplateSource: defaultGit,
+				ParameterSource: gitopsv1alpha1.GitConfig{
+					URI: defaultGit.URI,
+					Ref: defaultGit.Ref,
+				},
```
Comment (akavel) [inline pkg/handler/handler_test.go]: I think you can use `ParameterSource: defaultGit` here?

Comment (rhalat) [reply pkg/handler/handler_test.go]: Yes, thank you!

File: pkg/handler/handler_test.go (hunk 2 above)
Comment (rhalat) [reply pkg/handler/handler_test.go]: Done

File: pkg/handler/handler.go [hunk 4]
```diff
@@ -117,8 +117,11 @@ func WebhookHandler(w http.ResponseWriter, r *http.Request, reconciler gitopscon
...
+	return (strings.Contains(instance.Spec.TemplateSource.URI, *event.Repo.FullName) &&
+				 instance.Spec.TemplateSource.Ref == strings.TrimPrefix(*event.Ref, "refs/heads/")) ||
```
[the remark this answers is not in the document]
Comment (rhalat) [reply pkg/handler/handler.go]: There is no much information about that. All what I could find is webhook payload example:
https://developer.github.com/v3/activity/events/types/#webhook-payload-example-31

and topic from 2 weeks ago in GitHub community that it's not well documented:
https://github.community/t5/GitHub-API-Development-and/Push-webhook-payload-is-inexplicable/td-p/45416

But honestly I don't think that either `Ref` or `FullName` can be `nil`

[the remark this answers is not in the document]
Comment (rhalat) [reply pkg/handler/handler.go]: I added code to check if `event.Repo`, `event.Repo.FullName` and `event.Ref` are not `nil` in latest commit

File: pkg/handler/handler.go [hunk 5]
```diff
@@ -118,8 +118,9 @@ func WebhookHandler(w http.ResponseWriter, r *http.Request, reconciler gitopscon
 }
 
 func repoURLAndRefMatch(instance *gitopsv1alpha1.GitOpsConfig, event *github.PushEvent) bool {
-	return (strings.Contains(instance.Spec.TemplateSource.URI, *event.Repo.FullName) &&
-		instance.Spec.TemplateSource.Ref == strings.TrimPrefix(*event.Ref, "refs/heads/")) ||
+	return event.Repo != nil && event.Repo.FullName != nil && event.Ref != nil &&
+		(strings.Contains(instance.Spec.TemplateSource.URI, *event.Repo.FullName) &&
```
Comment (akavel) [inline pkg/handler/handler.go]: I think the `||` block needs to be in an extra pair of parentheses...

NIL_CHECKS && ((CONTAINS_CHECK && REF_CHECK) || (CONTAINS_CHECK && REF_CHECK))

otherwise I think the precedence rules will group all `&&`s together

Comment (rhalat) [reply pkg/handler/handler.go]: Fixed

**Comment by akavel** [comment]
@rhalat Please add an extra commit increasing the timeout in e2e-test.sh for hello-world-yaml-cr1. Please make sure the commit  message explains the reasons: from CI logs, we see that one run barely completes in time:

https://travis-ci.com/KohlsTechnology/eunomia/jobs/288138640#L2471

so the failure in the other run will probably disappear if we increase timeout:

https://travis-ci.com/KohlsTechnology/eunomia/jobs/288138641#L2469

As a side note, personally, I hate that the tests are taking so long now; I hope we'll find some way(s) to speed them up at some point.
